nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,78 @@
-
-At that point it makes more sense to break down the contracts into multiple assemblies. A possible structure solution is to have multiple assemblies per endpoint with a naming convention similar to:
-
-* EndpointName.Commands
-* EndpointName.Events
-
-Or even more strict:
-
-* EndpointName.AutonomousComponent.Commands
-* EndpointName.AutonomousComponent.Events
-
-While it might make sense to directly reference the \*.Commands assemblies from the sender and receiver endpoints, it might not for the \*.Events assemblies as they are used by many subscribing endpoints. In those cases, considering sharing the assembly through NuGet packages as opposed to using a direct reference.
+---
+title: Evolving Message Contracts
+summary: Guidelines for choosing a strategy for evolving message contracts
+reviewed: 2020-06-16
+component: Core
+isLearningPath: true
+---
+
+In message-based systems, the messages are part of the contract, which defines how services communicate with each other.
+
+Evolving contracts over time is not an easy task and the appropriate strategy should be defined individually for each system. When evolving message contracts, the following difficulties should be considered:
+
+* Endpoints updated to the latest message contract might still receive messages using the old contract since senders might still use the old contract or not all messages in queue have been processed before the upgrade.
+* Endpoints updated to the latest message contract might send messages using the new contract to endpoints that are still deployed with the old contract version.
+
+Generally the problem can't be resolved at the infrastructure level; therefore NServiceBus users must analyze their individual systems, consider how they are expected to evolve, and define the strategy which will make the most sense in their particular circumstances.
+
+This article presents basic guidelines for choosing contracts evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.
+
+TODO: link to new document
+Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).
+
+## Adding new message contracts
+
+When adding new message contracts to a contracts assembly, ensure to update the endpoints receveiving or subscribing to the new contract type first. When updating the sender/publisher first, the messages fail to be processed by endpoints that are not using the new contract assembly.
+
+## Adding data to existing contracts
+
+Adding additional data to existing contracts is the most common change of contracts. There are differnt approaches available.
+
+### Create a new contract type

```

Comment:
depends on what you'd consider "versioning". It's not assembly/package versioning (that can be used to indicate the change in the contract) but it would be something like this:

```
class OrderCreated : IMessage {
  string MyProperty1 {get;set;}
}

class OrderCreatedV2 : IEvent {
  string MyProperty1 {get;set;}
  string MyProperty2 {get;set;}
}
```

Reply:
(duck typing)